[MRG] Updated dbscan(): added documentation

Don86 opened the pull request:
Issue #8003 DBSCAN seems not to use multiple processors (n_jobs argument ignored).
Added documentation in dbscan() to alert users to missing functionality, and some information about why this functionality isn't always there.

Updated dbscan() documentation about missing functionality in n_jobs parameter.

Don86 committed:
Updated dbscan(): added documentation

Updated dbscan() documentation about missing functionality in n_jobs parameter.

Don86 committed:
Updated dbscan(): fixed whitespace

Fixed trailing whitespace at line 77.

Don86 committed:
Updated dbscan(): fixed whitespace (again)

Fixed whitespace again.

Don86 committed:
Update dbscan(): Take 3

Fixed whitespace at line 100.

jnothman commented:
nearest neighbor is parallelised in the brute case.

Don86 committed:
Update dbscan(): Addition for "brute" algorithm

Added "Setting algorithm="brute" uses multiple cores, but may cause a slow down instead." at line 103.

jnothman commented on sklearn/cluster/dbscan_.py:
```diff
@@ -100,7 +100,8 @@ def dbscan(X, eps=0.5, min_samples=5, metric='minkowski',
     NearestNeighbour, passing the n_jobs parameter to it, but NearestNeighbour
-    has not yet been parrallelized.
+    has not yet been parrallelized. Setting algorithm="brute" uses multiple
+    cores, but may cause a slow down instead.
```
all parallelism may cause a slow down

Don86 replied:
Changed wording to reflect this: changed "may" to "will", removed "instead".

jnothman commented on sklearn/cluster/dbscan_.py:
```diff
@@ -100,7 +100,8 @@ def dbscan(X, eps=0.5, min_samples=5, metric='minkowski',
     Note about n_jobs: this does not seem to use multiple processors. Calls
     NearestNeighbour, passing the n_jobs parameter to it, but NearestNeighbour
-    has not yet been parrallelized.
+    has not yet been parrallelized. Setting algorithm="brute" uses multiple
```
I only think we ever intended to parallelise the nearest neighbors.

Don86 replied:
I assume you mean: "I *don't* think we ever intended to parallelize the nearest neighbors"?
Removed "yet".

jnothman replied:
No, I meant what I said: I don't think we had any intentions to parallelise dbscan beyond its nn queries.

Don86 committed:
Updated dbscan(): update 5

Updated comments in response to latest feedback.

tguillemot commented on sklearn/cluster/dbscan_.py:
```diff
@@ -96,6 +98,11 @@ def dbscan(X, eps=0.5, min_samples=5, metric='minkowski',
+    Note about n_jobs: this does not seem to use multiple processors. Calls
+    NearestNeighbour, passing the n_jobs parameter to it, but NearestNeighbour
+    has not been parrallelized. Setting algorithm="brute" uses multiple
+    cores, but will cause a slow down.
```
What do you think of :

Note about \`\`n_jobs\`\`: when \`\`algorithm="brute"\`\`, the \`\`n_jobs\`\` parameter is used to compute a brute force multiple processors approach which can cause a slow down. For a faster computation we recommend to use other NearestNeighbors algorithms by choosing another value for the parameter \`\`algorithm\`\`. Nevertheless, these  NearestNeighbors methods are not parrellelized in scikit-learn and will not use the \`\`n_jobs\`\` parameter.

Don86 replied:
Hi, thanks for your feedback. Yup, sounds good, I've made the changes.

tguillemot commented on sklearn/cluster/dbscan_.py:
```diff
@@ -74,6 +74,8 @@ def dbscan(X, eps=0.5, min_samples=5, metric='minkowski',
         The number of parallel jobs to run for neighbors search.
         If ``-1``, then the number of jobs is set to the number of CPU cores.
+        Currently may not work as expected; might not use multiple processors.
+        See notes below.
```
Maybe it's better to directly add the note information here.

tguillemot commented:
Thanks @Don86. I've done some suggestions, tell me what you think ?

Don86 committed:
Update dbscan()

Rewrite in response to feedback.

jnothman commented on sklearn/cluster/dbscan_.py:
```diff
@@ -74,6 +74,13 @@ def dbscan(X, eps=0.5, min_samples=5, metric='minkowski',
     n_jobs : int, optional (default = 1)
         The number of parallel jobs to run for neighbors search.
         If ``-1``, then the number of jobs is set to the number of CPU cores.
+        Currently may not work as expected. When ``algorithm="brute"``, the 
```
How about "Parallelism is only currently used when algorithm="brute", though other algorithms may remain faster."

Don86 committed:
Updated dbscan(): Update 8

Updated in response to feedback
Fixed whitespace

Don86 committed:
Update fit_predict(): documentation

Changed line 151 to "Returns 1 for inliers and -1 for anomalies/outliers." to match main description; peviously mismatched.

jnothman commented on sklearn/cluster/dbscan_.py:
```diff
@@ -74,6 +74,13 @@ def dbscan(X, eps=0.5, min_samples=5, metric='minkowski',
     n_jobs : int, optional (default = 1)
         The number of parallel jobs to run for neighbors search.
         If ``-1``, then the number of jobs is set to the number of CPU cores.
+        Parrallelism currently only used when ``algorithm="brute"``, where
```
I'm really not finding this clear and succinct enough, which is why I suggested a specific wording.

jnothman commented on sklearn/neighbors/lof.py:
```diff
@@ -148,7 +148,7 @@ def fit_predict(self, X, y=None):
         -------
         is_inlier : array, shape (n_samples,)
-            Returns 1 for anomalies/outliers and -1 for inliers.
+            Returns 1 for inliers and -1 for anomalies/outliers.
```
Doesn't belong here

Don86 committed:
Update fit_predict()

Rolled back some changes which don't belong in this branch.

tguillemot commented on sklearn/neighbors/lof.py:
```diff
@@ -148,7 +148,7 @@ def fit_predict(self, X, y=None):
         -------
         is_inlier : array, shape (n_samples,)
-            Returns 1 for inliers and -1 for anomalies/outliers.
+            Returns 1 for anomalies/outliers and -1 for inliers.
```
Is it related to dbscan ? If it is not, can you open another PR. 
BTW, it seems that this confusion is done in a lot of places in the LOF file (ex : [l138](https://github.com/scikit-learn/scikit-learn/pull/8039/files/176979e3f794eeece81a1ae60e1925a97b2e26ec..1b8c8c259f4006bf60cfd961529da60fb14b717c#diff-4985428b0bff6aa71f59fed1f8a72f93R138) ). 
Can you check the entire file ?

tguillemot replied:
Sorry for the my mistake. 
I had a look to the diff not the PR.

amueller commented:
@jnothman is this still relevant?

jnothman reviewed (commented):
Not really relevant anymore. It is now only ineffective when metric='precomputed'
